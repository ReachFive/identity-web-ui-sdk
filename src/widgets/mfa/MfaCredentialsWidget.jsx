import React from 'react';

import { createMultiViewWidget } from '../../components/widget/widget';

import {simpleField} from '../../components/form/fields/simpleField';
import {Info, Intro, Separator} from '../../components/miscComponent';
import {createForm} from '../../components/form/formComponent';
import {deepDefaults} from '../../helpers/deepDefaults';
import phoneNumberField from '../../components/form/fields/phoneNumberField';
import {withTheme} from '../../components/widget/widgetContext';
import styled from 'styled-components';


const EmailRegisteringCredentialForm = createForm({
    prefix: 'r5-mfa-credentials-email-',
    submitLabel: 'mfa.register.email'
});

const EmailCredentialRemovalForm = createForm({
    prefix: 'r5-mfa-credentials-email-removal-',
    submitLabel: 'mfa.remove.email'
})

const PhoneNumberCredentialRemovalForm = config => createForm({
    prefix: 'r5-mfa-credentials-phone-number-removal-',
    fields: [
        phoneNumberField({required: true}, config)
    ],
    submitLabel: 'mfa.remove.phoneNumber'
})

const VerificationCodeForm = createForm({
    prefix: 'r5-mfa-credentials-verification-code-',
    fields: [
        simpleField({
            key: 'verification_code',
            label: 'verificationCode',
            type: 'text'
        })
    ],
})

const PhoneNumberRegisteringCredentialForm = config => createForm({
    prefix: 'r5-mfa-credentials-phone-number-',
    fields: [
        phoneNumberField({required: true}, config)
    ],
    submitLabel: 'mfa.register.phoneNumber'
})

const DivCredentialBlock = withTheme(styled.div`
    margin-left: ${props => props.theme.get('_blockInnerHeight')}px;
    margin-bottom: 5em;
`);

class MainView extends React.Component {
    onEmailRegistering = _ => {
        return this.props.apiClient.startMfaEmailRegistration({
                accessToken: this.props.accessToken
            }
        )
    }

    onPhoneNumberRegistering = data => {
        return this.props.apiClient.startMfaPhoneNumberRegistration({
            accessToken: this.props.accessToken,
            ...data
        })
    }

    onEmailRemoval = _ => {
        return this.props.apiClient.removeMfaEmail({
            accessToken: this.props.accessToken
        })
    }

    onPhoneNumberRemoval = data => {
        return this.props.apiClient.removeMfaPhoneNumber({
            accessToken: this.props.accessToken,
            phoneNumber: data.phoneNumber
        })
    }

    render() {
        const { i18n, showIntro, config, showRemoveMfaCredentials } = this.props
        const PhoneNumberInputForm = PhoneNumberRegisteringCredentialForm(config);
        const PhoneNumberRemovalInputForm = PhoneNumberCredentialRemovalForm(config)
        return (
            <div>
                <DivCredentialBlock>
                    {config.mfaEmailEnabled &&
                        <div>
                            {showIntro && <Intro>{i18n('mfa.email.explain')}</Intro>}
                            <EmailRegisteringCredentialForm handler={this.onEmailRegistering} onSuccess={data => this.props.goTo('verification-code', {...data, registrationType: 'email'})}/>
                        </div>
                    }

                    {config.mfaEmailEnabled && config.mfaSmsEnabled && <Separator/>}
                    {config.mfaSmsEnabled &&
                        <div>
                            {showIntro && <Intro>{i18n('mfa.phoneNumber.explain')}</Intro>}
                            <PhoneNumberInputForm
                                handler={this.onPhoneNumberRegistering} onSuccess={data => this.props.goTo('verification-code', {...data, registrationType: 'sms'})}/>
                        </div>
                    }
                </DivCredentialBlock>
                <DivCredentialBlock>
                    {showRemoveMfaCredentials &&
                        config.mfaEmailEnabled &&
                        <div>
                            {showIntro && <Intro>{i18n('mfa.email.remove.explain')}</Intro>}
                            <EmailCredentialRemovalForm handler={this.onEmailRemoval}
                                                        onSuccess={_ => this.props.goTo('credential-removed', {credentialType: 'email'})}/>
                        </div>
                    }
                    {showRemoveMfaCredentials && config.mfaEmailEnabled && config.mfaSmsEnabled && <Separator/>}
                    {showRemoveMfaCredentials &&
                        config.mfaSmsEnabled &&
                        <div>
                            {showIntro && <Intro>{i18n('mfa.phoneNumber.remove.explain')}</Intro>}
                            <PhoneNumberRemovalInputForm handler={this.onPhoneNumberRemoval}
                                                         onSuccess={_ => this.props.goTo('credential-removed', {credentialType: 'sms'})}/>
                        </div>
                    }
                </DivCredentialBlock>
            </div>
        )
    }
}

class VerificationCodeView extends React.Component {
    onEmailCodeVerification = data => {
        return this.props.apiClient.verifyMfaEmailRegistration({
            ...data,
            accessToken: this.props.accessToken
        })
    }

    onSmsCodeVerification = data => {
        return this.props.apiClient.verifyMfaPhoneNumberRegistration({
            ...data,
            accessToken: this.props.accessToken
        })
    }

    render() {
        const { i18n, showIntro, status, goTo, registrationType } = this.props
        return <div>
            {showIntro && status === 'email_sent' && <Intro>{i18n('mfa.verify.email')}</Intro>}
            {status === 'email_sent' && <VerificationCodeForm handler={this.onEmailCodeVerification} onSuccess={data => goTo('credential-registered', {...data, registrationType})}/>}

            {showIntro && status === 'sms_sent' && <Intro>{i18n('mfa.verify.sms')}</Intro>}
            {status === 'sms_sent' && <VerificationCodeForm handler={this.onSmsCodeVerification} onSuccess={data => goTo('credential-registered', {...data, registrationType})}/>}

            {showIntro && status === 'enabled' && goTo('credential-registered', {i18n, registrationType})}
        </div>
    }
}

const CredentialRegisteredView = ({ i18n, registrationType}) => <div>
        {registrationType === 'email' && <Info>{i18n('mfa.email.registered')}</Info>}
        {registrationType === 'sms' && <Info>{i18n('mfa.phoneNumber.registered')}</Info>}
    </div>

const CredentialRemovedView = ({ i18n, credentialType}) => <div>
    {credentialType === 'email' && <Info>{i18n('mfa.email.removed')}</Info>}
    {credentialType === 'sms' && <Info>{i18n('mfa.phoneNumber.removed')}</Info>}
</div>

export default createMultiViewWidget({
    initialView: 'main',
    views: {
        'main': MainView,
        'credential-registered': CredentialRegisteredView,
        'verification-code': VerificationCodeView,
        'credential-removed': CredentialRemovedView
    },
    prepare: (options) => {
        return deepDefaults({
            showIntro: true,
            showRemoveMfaCredentials: true,
            ...options
        }
    )}
})
