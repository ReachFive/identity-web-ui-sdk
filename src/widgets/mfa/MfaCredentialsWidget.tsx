import React from 'react';
import styled from 'styled-components';
import { MFA } from '@reachfive/identity-core';
import type { StartMfaEmailRegistrationResponse, StartMfaPhoneNumberRegistrationResponse } from '@reachfive/identity-core/es/main/mfaClient';

import type { Config, Prettify } from '../../types';

import { createMultiViewWidget } from '../../components/widget/widget';

import { simpleField } from '../../components/form/fields/simpleField';
import { Info, Intro, Separator } from '../../components/miscComponent';
import { createForm } from '../../components/form/formComponent';
import phoneNumberField, { type PhoneNumberOptions } from '../../components/form/fields/phoneNumberField';

import { UserError } from '../../helpers/errors';

import { useI18n } from '../../contexts/i18n';
import { useReachfive } from '../../contexts/reachfive';
import { useRouting } from '../../contexts/routing';
import { useConfig } from '../../contexts/config';


const EmailRegisteringCredentialForm = createForm({
    prefix: 'r5-mfa-credentials-email-',
    submitLabel: 'mfa.register.email'
});

const EmailCredentialRemovalForm = createForm({
    prefix: 'r5-mfa-credentials-email-removal-',
    submitLabel: 'mfa.remove.email'
})

const PhoneNumberCredentialRemovalForm = createForm({
    prefix: 'r5-mfa-credentials-phone-number-removal-',
    submitLabel: 'mfa.remove.phoneNumber'
})

type VerificationCodeFormData = { verificationCode: string }

const VerificationCodeForm = createForm<VerificationCodeFormData>({
    prefix: 'r5-mfa-credentials-verification-code-',
    fields: [
        simpleField({
            key: 'verification_code',
            label: 'verificationCode',
            type: 'text'
        })
    ],
})

type PhoneNumberRegisteringCredentialFormData = { phoneNumber: string }

const PhoneNumberRegisteringCredentialForm = (config: Config) => createForm<PhoneNumberRegisteringCredentialFormData, { phoneNumberOptions?: PhoneNumberOptions }>({
    prefix: 'r5-mfa-credentials-phone-number-',
    fields: ({ phoneNumberOptions }) => ([
        phoneNumberField({
            required: true,
            ...phoneNumberOptions,
        }, config)
    ]),
    submitLabel: 'mfa.register.phoneNumber'
})

const DivCredentialBlock = styled.div`
    margin-left: ${props => props.theme._blockInnerHeight}px;
    margin-bottom: 5em;
`;

interface MainViewProps {
    /**
     * The authorization credential JSON Web Token (JWT) used to access the ReachFive API, less than five minutes old.
     */
    accessToken: string
    /**
     * The userâ€™s MFA credentials
     */
    credentials: MFA.CredentialsResponse['credentials']
    /**
     * Boolean to enable (`true`) or disable (`false`) whether the option to remove MFA credentials are displayed.
     * 
     * @default false
     */
    requireMfaRegistration?: boolean
    /**
     * Show the introduction text.
     * 
     * @default true
     */
    showIntro?: boolean
    /**
     * Boolean to enable (true) or disable (false) whether the option to remove MFA credentials are displayed.
     * 
     * @default true
     */
    showRemoveMfaCredentials?: boolean
    /**
     * Phone number field options.
     */
    phoneNumberOptions?: PhoneNumberOptions
    /**
     * Callback function called when the request has failed.
     */
    onError?: (error: unknown) => void
}

const MainView = ({
    accessToken,
    credentials,
    onError = () => {},
    phoneNumberOptions,
    requireMfaRegistration = false,
    showIntro = true,
    showRemoveMfaCredentials = true,
}: MainViewProps) => {
    const coreClient = useReachfive()
    const config = useConfig()
    const i18n = useI18n()
    const { goTo } = useRouting()

    const onEmailRegistering = () => {
        return coreClient
            .startMfaEmailRegistration({
                    accessToken
                }
            )
            .catch(error => {
                onError(error)
                return Promise.reject(error)
            })
    }

    const onPhoneNumberRegistering = (data: PhoneNumberRegisteringCredentialFormData) => {
        return coreClient
            .startMfaPhoneNumberRegistration({
                accessToken,
                ...data
            })
            .catch(error => {
                onError(error)
                return Promise.reject(error)
            })
    }

    const onEmailRemoval = () => {
        return coreClient
            .removeMfaEmail({
                accessToken
            })
            .catch(error => {
                onError(error)
                return Promise.reject(error)
            })
    }

    const onPhoneNumberRemoval = ({ phoneNumber }: MFA.PhoneCredential) => {
        return coreClient
            .removeMfaPhoneNumber({
                accessToken,
                phoneNumber,
            })
            .catch(onError)
    }

    const PhoneNumberInputForm = PhoneNumberRegisteringCredentialForm(config);
    const phoneNumberCredentialRegistered = credentials.find<MFA.PhoneCredential>(
        (credential): credential is MFA.PhoneCredential => MFA.isPhoneCredential(credential)
    )
    const isEmailCredentialRegistered = credentials.some(credential => MFA.isEmailCredential(credential))
    const isPhoneCredentialRegistered = credentials.some(credential => MFA.isPhoneCredential(credential))

    return (
        <div>
            <DivCredentialBlock>
                {config.mfaEmailEnabled && !isEmailCredentialRegistered &&
                    <div>
                        {showIntro && <Intro>{requireMfaRegistration ? i18n('mfa.email.explain.required') :i18n('mfa.email.explain')}</Intro>}
                        <EmailRegisteringCredentialForm
                            handler={onEmailRegistering}
                            onSuccess={(data: StartMfaEmailRegistrationResponse) => goTo<VerificationCodeViewState>('verification-code', {...data, registrationType: 'email'})}
                        />
                    </div>
                }

                {config.mfaEmailEnabled && config.mfaSmsEnabled && !isPhoneCredentialRegistered && <Separator text={i18n('or')} />}

                {config.mfaSmsEnabled && !isPhoneCredentialRegistered &&
                    <div>
                        {showIntro && <Intro>{i18n('mfa.phoneNumber.explain')}</Intro>}
                        <PhoneNumberInputForm
                            handler={onPhoneNumberRegistering}
                            onSuccess={(data: StartMfaPhoneNumberRegistrationResponse) => goTo<VerificationCodeViewState>('verification-code', {...data, registrationType: 'sms'})}
                            phoneNumberOptions={phoneNumberOptions}
                        />
                    </div>
                }
            </DivCredentialBlock>
            <DivCredentialBlock>
                {showRemoveMfaCredentials &&
                    config.mfaEmailEnabled &&
                    isEmailCredentialRegistered &&
                    <div>
                        {showIntro && <Intro>{i18n('mfa.email.remove.explain')}</Intro>}
                        <EmailCredentialRemovalForm
                            handler={onEmailRemoval}
                            onSuccess={() => goTo<CredentialRemovedViewState>('credential-removed', { credentialType: 'email' })}
                        />
                    </div>
                }
                {showRemoveMfaCredentials 
                    && config.mfaEmailEnabled 
                    && config.mfaSmsEnabled 
                    && phoneNumberCredentialRegistered 
                    && isEmailCredentialRegistered 
                    && <Separator text={i18n('or')} />
                }
                {showRemoveMfaCredentials &&
                    config.mfaSmsEnabled &&
                    phoneNumberCredentialRegistered &&
                    <div>
                        {showIntro && <Intro>{i18n('mfa.phoneNumber.remove.explain')}</Intro>}
                        <PhoneNumberCredentialRemovalForm
                            handler={() => onPhoneNumberRemoval({ ...phoneNumberCredentialRegistered })}
                            onSuccess={() => goTo<CredentialRemovedViewState>('credential-removed', {credentialType: 'sms'})}
                        />
                    </div>
                }
            </DivCredentialBlock>
        </div>
    )
}

interface VerificationCodeViewProps {
    /**
     * The authorization credential JSON Web Token (JWT) used to access the ReachFive API, less than five minutes old.
     */
    accessToken: string
    /**
     * Show the introduction text.
     */
    showIntro?: boolean
    /**
     * Callback function called when the request has succeed.
     */
    onSuccess?: () => void
    /**
     * Callback function called when the request has failed.
     */
    onError?: (error: unknown) => void
}

type VerificationCodeViewState = 
    ({ registrationType: 'email' } & StartMfaEmailRegistrationResponse) | 
    ({ registrationType: 'sms' } & StartMfaPhoneNumberRegistrationResponse)

const VerificationCodeView = ({
    accessToken,
    onError = () => {},
    onSuccess = () => {},
    showIntro = true
}: VerificationCodeViewProps) => {
    const coreClient = useReachfive()
    const i18n = useI18n()
    const { goTo, params } = useRouting()
    const { registrationType, status } = params as VerificationCodeViewState

    const onEmailCodeVerification = (data: VerificationCodeFormData) => {
        return coreClient
            .verifyMfaEmailRegistration({
                ...data,
                accessToken
            })
            .catch(error => {
                onError(error)
                return Promise.reject(error)
            })
    }

    const onSmsCodeVerification = (data: VerificationCodeFormData) => {
        return coreClient
            .verifyMfaPhoneNumberRegistration({
                ...data,
                accessToken
            })
            .catch(error => {
                onError(error)
                return Promise.reject(error)
            })
    }

    const onCredentialRegistered = () => {
        onSuccess()
        goTo<CredentialRegisteredViewState>('credential-registered', { registrationType })
    }

    if (showIntro && status === 'enabled') {
        goTo<CredentialRegisteredViewState>('credential-registered', { registrationType })
        return null
    }

    return (
        <div>
            {showIntro && status === 'email_sent' && <Intro>{i18n('mfa.verify.email')}</Intro>}
            {status === 'email_sent' &&
                <VerificationCodeForm
                    handler={onEmailCodeVerification}
                    onSuccess={onCredentialRegistered}
                />
            }

            {showIntro && status === 'sms_sent' && <Intro>{i18n('mfa.verify.sms')}</Intro>}
            {status === 'sms_sent' &&
                <VerificationCodeForm
                    handler={onSmsCodeVerification}
                    onSuccess={onCredentialRegistered}
                />
            }
        </div>
    )
}

interface CredentialRegisteredViewProps {}

type CredentialRegisteredViewState = {
    registrationType: MFA.CredentialsResponse['credentials'][number]['type']
}

const CredentialRegisteredView = () => {
    const i18n = useI18n()
    const { params } = useRouting()
    const { registrationType } = params as CredentialRegisteredViewState
    return (
        <div>
            {registrationType === 'email' && <Info>{i18n('mfa.email.registered')}</Info>}
            {registrationType === 'sms' && <Info>{i18n('mfa.phoneNumber.registered')}</Info>}
        </div>
    )
}

type CredentialRemovedViewProps = {}

type CredentialRemovedViewState = {
    credentialType: MFA.CredentialsResponse['credentials'][number]['type']
}

const CredentialRemovedView = () => {
    const i18n = useI18n()
    const { params } = useRouting()
    const { credentialType } = params as CredentialRemovedViewState
    return (
        <div>
            {credentialType === 'email' && <Info>{i18n('mfa.email.removed')}</Info>}
            {credentialType === 'sms' && <Info>{i18n('mfa.phoneNumber.removed')}</Info>}
        </div>
    )
}

type MfaCredentialsProps = Prettify<MainViewProps & CredentialRegisteredViewProps & VerificationCodeViewProps & CredentialRemovedViewProps>

export type MfaCredentialsWidgetProps = Prettify<Omit<MfaCredentialsProps, 'credentials'>>

export default createMultiViewWidget<MfaCredentialsWidgetProps, MfaCredentialsProps>({
    initialView: 'main',
    views: {
        'main': MainView,
        'credential-registered': CredentialRegisteredView,
        'verification-code': VerificationCodeView,
        'credential-removed': CredentialRemovedView
    },
    prepare: (options, { apiClient }) => {
        return apiClient.listMfaCredentials(options.accessToken)
            .catch(error => {
                options.onError?.(error)
                throw UserError.fromAppError(error)
            })
            .then(({ credentials }) => ({
                ...options,
                credentials,
            }))
    }
})
