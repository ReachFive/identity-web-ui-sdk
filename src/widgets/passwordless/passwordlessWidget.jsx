import React from 'react';

import { email } from '../../core/validation';

import { createMultiViewWidget } from '../../components/widget/widget';
import { Info, Intro, Separator } from '../../components/miscComponent';

import { createForm } from '../../components/form/formComponent';
import { simpleField } from '../../components/form/fields/simpleField';
import phoneNumberField from '../../components/form/fields/phoneNumberField';
import SocialButtons from '../../components/form/socialButtonsComponent';

const EmailInputForm = createForm({
    prefix: 'r5-passwordless-',
    fields: [
        simpleField({
            key: 'email',
            label: 'email',
            type: 'email',
            validator: email
        })
    ]
});

const phoneNumberInputForm = config => createForm({
    prefix: 'r5-passwordless-sms-',
    fields: [phoneNumberField({ required: true }, config)]
});

const VerificationCodeInputForm = createForm({
    prefix: 'r5-passwordless-sms-',
    fields: [
        simpleField({
            key: 'verification_code',
            label: 'verificationCode',
            type: 'text'
        })
    ]
});

class MainView extends React.Component {
    handleSubmit = data => this.props.apiClient.startPasswordless(data, this.props.auth).then(_ => data);

    handleSuccess = data => {
        return data.email
            ? this.props.goTo('emailSent')
            : this.props.goTo('verificationCode', data);
    };

    render() {
        const { i18n, showSocialLogins, socialProviders, showIntro, authType } = this.props;
        const isEmail = authType === 'magic_link' || authType === '' || !authType;
        const PhoneNumberInputForm = phoneNumberInputForm(this.props.config);

        return <div>
            {showSocialLogins && socialProviders && socialProviders.length > 0 && (
                <SocialButtons providers={socialProviders} auth={this.props.auth} />
            )}
            {showSocialLogins && socialProviders && socialProviders.length > 0 && (
                <Separator text={i18n('or')} />
            )}
            {isEmail && showIntro && <Intro>{i18n('passwordless.intro')}</Intro>}
            {isEmail && <EmailInputForm handler={this.handleSubmit} onSuccess={this.handleSuccess} />}
            {!isEmail && showIntro && <Intro>{i18n('passwordless.sms.intro')}</Intro>}
            {!isEmail && <PhoneNumberInputForm handler={this.handleSubmit} onSuccess={this.handleSuccess} />}
        </div>;
    }
}

class VerificationCodeView extends React.Component {
    handleSubmit = data => {
        const { apiClient, auth, phoneNumber } = this.props;

        return apiClient.verifyPasswordless({ phoneNumber, ...data }, auth);
    };

    render() {
        return <div>
            <Info>{this.props.i18n('passwordless.sms.verification.intro')}</Info>
            <VerificationCodeInputForm handler={this.handleSubmit} />
        </div>;
    }
}
const EmailSentView = ({ i18n }) => <Info>{i18n('passwordless.emailSent')}</Info>;

export default createMultiViewWidget({
    initialView: 'main',
    views: {
        main: MainView,
        emailSent: EmailSentView,
        verificationCode: VerificationCodeView
    },
    prepare: (options, { config }) => ({
        socialProviders: config.socialProviders,
        auth: {},
        showSocialLogins: false,
        showIntro: true,
        authType: 'magic_link',
        ...options
    })
})
