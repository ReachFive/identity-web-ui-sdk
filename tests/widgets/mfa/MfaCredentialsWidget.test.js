import Enzyme, {render} from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import passwordlessWidget from "../../../src/widgets/passwordless/passwordlessWidget";
import renderer from "react-test-renderer";
import MfaCredentialsWidget from "../../../src/widgets/mfa/MfaCredentialsWidget";
import $ from "cheerio";


Enzyme.configure({ adapter: new Adapter() });
const defaultConfig = { domain: 'local.reach5.net', mfaEmailEnabled: true, mfaSmsEnabled: true };
const textFilter = expected => (i, el) => $(el).text() === expected;

describe('Snapshot', () => {
    const generateSnapshot = (options, config = defaultConfig) => () => {
        const tree = MfaCredentialsWidget(options, { config, apiClient: {} })
            .then(result => renderer.create(result).toJSON());

        expect(tree).resolves.toMatchSnapshot();
    };

    describe('mfaCredentials', () => {
        test('default', generateSnapshot());

        test('no intro', generateSnapshot({ showIntro: false }));
    });
});
describe('DOM testing', () => {
    const generateComponent = async (options, config = defaultConfig) => {
        const result = await MfaCredentialsWidget(options, { config, apiClient: {} });

        return render(result);
    };

    describe('mfaCredentials', () => {
        test('default', async () => {
            const instance = await generateComponent({});
            // Intro
            expect(
                instance.find('div').filter(textFilter('mfa.email.explain'))
            ).toHaveLength(1);

            // Form button sms
            expect(
                instance.find('button').filter(textFilter('mfa.register.phoneNumber'))
            ).toHaveLength(1);

            // Form button email
            expect(
                instance.find('button').filter(textFilter('mfa.register.email'))
            ).toHaveLength(1);
        });
    });
});
