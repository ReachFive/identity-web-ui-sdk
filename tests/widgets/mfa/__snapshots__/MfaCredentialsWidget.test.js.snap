// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot mfaCredentials default 1`] = `
<div>
  <div
    class="sc-imWZod jkKkUW"
  >
    <div>
      <div
        class="sc-gFqAYk bccvP"
      >
        <div>
          <div
            class="sc-gEvDqW sc-eqUzNf juyGmH cnAvcW"
          >
            mfa.email.explain
          </div>
          <form
            class="sc-eldOKa kdtPOR"
            novalidate=""
          >
            <button
              class="sc-eDPFhE lHVpj"
              data-testid="submit"
              type="submit"
            >
              mfa.register.email
            </button>
          </form>
        </div>
        <div
          class="sc-iGgVNO kzlSaL"
        >
          <span>
            or
          </span>
        </div>
        <div>
          <div
            class="sc-gEvDqW sc-eqUzNf juyGmH cnAvcW"
          >
            mfa.phoneNumber.explain
          </div>
          <form
            class="sc-eldOKa kdtPOR"
            novalidate=""
          >
            <div
              class="sc-dAlxHm jtSANl"
            >
              <label
                class="sc-dhKdPU iqUOyk"
                for="phone_number"
              >
                phoneNumber
              </label>
              <input
                autocomplete="tel"
                class="sc-jlZhRR dylVxL"
                data-testid="phone_number"
                id="phone_number"
                labels="[object Object]"
                name="phone_number"
                placeholder="phoneNumber"
                required=""
                type="tel"
                value=""
              />
            </div>
            <button
              class="sc-eDPFhE lHVpj"
              data-testid="submit"
              type="submit"
            >
              mfa.register.phoneNumber
            </button>
          </form>
        </div>
      </div>
      <div
        class="sc-gFqAYk bccvP"
      />
    </div>
  </div>
</div>
`;

exports[`Snapshot mfaCredentials no intro 1`] = `
<div>
  <div
    class="sc-imWZod jkKkUW"
  >
    <div>
      <div
        class="sc-gFqAYk bccvP"
      />
      <div
        class="sc-gFqAYk bccvP"
      >
        <div>
          <form
            class="sc-eldOKa kdtPOR"
            novalidate=""
          >
            <button
              class="sc-eDPFhE lHVpj"
              data-testid="submit"
              type="submit"
            >
              mfa.remove.email
            </button>
          </form>
        </div>
        <div
          class="sc-iGgVNO kzlSaL"
        >
          <span>
            or
          </span>
        </div>
        <div>
          <form
            class="sc-eldOKa kdtPOR"
            novalidate=""
          >
            <button
              class="sc-eDPFhE lHVpj"
              data-testid="submit"
              type="submit"
            >
              mfa.remove.phoneNumber
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
